### PR TITLE
[p2p] Find nodes till maximum

### PR DESCRIPTION
**Pull request by minsulee2**
Updates:
- Add p2pStates (this is not states for each p2p node connection)
- Now automatically look for nodes till made fully connected
- Update unittest

Cannot somehow reproduce the flaky p2p unittest bug #580

### CODE REVIEW

File: p2p/index.js [hunk 1]
```diff
@@ -181,13 +173,52 @@ class P2pClient {
     this.trackerWebSocket.send(JSON.stringify(message));
   }
 
+  setIntervalForTrackerConnection() {
+    if (!this.intervalTrackerConnection) {
+      this.intervalTrackerConnection = setInterval(() => {
+        this.connectToTracker();
+      }, RECONNECT_INTERVAL_MS);
```
Comment (liayoo) [inline p2p/index.js:180]: how about TRACKER_RECONNECT_INTERVAL_MS, as we now have NODE_CONNECTION_INVERVAL_MS as well?

File: p2p/index.js [hunk 2]
```diff
@@ -181,13 +173,52 @@ class P2pClient {
...
       this.updatePeerInfoToTracker();
     }, UPDATE_TO_TRACKER_INTERVAL_MS);
   }
 
+  clearIntervalForTrackerUpdate() {
+    clearInterval(this.intervalTrackerUpdate);
+    this.intervalTrackerUpdate = null;
+  }
+
+  setIntervalForConnectingNodes() {
```
Comment (liayoo) [inline p2p/index.js:201]: how about setIntervalForNodeConnection? (like clearIntervalForNodeConnection)

**Review by liayoo — APPROVED** [review]
LGTM with some comments. Thanks!

File: common/constants.js [hunk 3]
```diff
@@ -160,6 +160,17 @@ const BlockchainNodeStates = {
   SERVING: 'SERVING',
 };
 
+/**
+ * States of p2p network.
+ *
+ * @enum {string}
+ */
+const p2pNetworkStates = {
```
Comment (platfowner) [inline common/constants.js:168]: What about 'P2pNetworkStates'?

File: p2p/index.js [hunk 4]
```diff
@@ -51,6 +53,8 @@ class P2pClient {
         this, node, minProtocolVersion, maxProtocolVersion, this.maxInbound);
     this.trackerWebSocket = null;
     this.outbound = {};
+    this.state = p2pNetworkStates.STARTING;
```
Comment (platfowner) [inline p2p/index.js:56]: What about 'p2pState'?

File: p2p/index.js (hunk 2 above)
Comment (platfowner) [reply p2p/index.js:201]: What about 'Peers' instead of 'Nodes'?

File: common/constants.js [hunk 5]
```diff
@@ -160,6 +160,17 @@ const BlockchainNodeStates = {
...
+  STARTING: 'STARTING',
+  CONNECTING: 'CONNECTING',
+  FULLY_SERVING: 'FULLY_SERVING'
```
Comment (platfowner) [inline common/constants.js:171]: minor: what about 'STARTING', 'UNSTEADY' (or 'EXPANDING' or 'DISCOVERING') and 'STEADY'?

Comment (minsulee2) [reply common/constants.js:171]: I'm loving it! I took STARTING, EXPANDING, and STEADY!

File: tracker-server/network-topology.js [hunk 6]
```diff
@@ -2,15 +2,19 @@ const commonUtil = require('../common/common-util');
 const { abbrAddr } = require('./util');
 
 const _buildGraphData = (peerNodes) => {
+  const filteredPeerNodesEntries = Object.entries(peerNodes).filter(([address, peerNode]) => {
+    return peerNode.isAlive === true
+  });
+  const alivePeerNodes = Object.fromEntries(filteredPeerNodesEntries);
```
[the remark this answers is not in the document]
Comment (minsulee2) [reply tracker-server/network-topology.js:8]: Done!

File: p2p/index.js (hunk 1 above)
Comment (minsulee2) [reply p2p/index.js]: Thanks! all conventions are great now !

File: p2p/index.js [hunk 7]
```diff
@@ -181,13 +173,52 @@ class P2pClient {
...
+    clearInterval(this.intervalTrackerUpdate);
+    this.intervalTrackerUpdate = null;
+  }
+
+  setIntervalForConnectingNodes() {
+    if (Object.keys(this.outbound).length < TARGET_NUM_OUTBOUND_CONNECTION) {
+      this.state = p2pNetworkStates.CONNECTING;
+      this.sendRequestForNewPeers();
+    } else {
+      this.state = p2pNetworkStates.FULLY_SERVING;
```
[the remark this answers is not in the document]
Comment (minsulee2) [reply p2p/index.js]: I took @platfowner 's below! Now, STARTING, EXPANDING, and STEADY!

File: p2p/index.js [hunk 8]
```diff
@@ -181,11 +173,49 @@ class P2pClient {
...
+  }
+
+  clearIntervalForTrackerUpdate() {
+    clearInterval(this.intervalTrackerUpdate);
+    this.intervalTrackerUpdate = null;
+  }
+
+  updateP2pState() {
+    if (Object.keys(this.outbound).length < TARGET_NUM_OUTBOUND_CONNECTION) {
+      this.p2pState = P2pNetworkStates.CONNECTING;
```
Comment (platfowner) [inline p2p/index.js:203]: 'EXPANDING'?

File: p2p/index.js [hunk 9]
```diff
@@ -181,11 +173,49 @@ class P2pClient {
...
+
+  setIntervalForPeerConnection() {
+    this.updateP2pState();
+    this.intervalNodeConnection = setInterval(() => {
+      this.updateP2pState();
+    }, NODE_CONNECTION_INVERVAL_MS);
+  }
+
+  clearIntervalForPeerConnection() {
+    clearInterval(this.intervalNodeConnection);
```
[the remark this answers is not in the document]
Comment (minsulee2) [reply p2p/index.js:218]: Done!

**Pull request merged**